src/dispatch/case/flows.py

```diff
@@ -17,18 +19,130 @@
+                case_id=case.id,
+            )
+
+    return individual_contacts, team_contacts
+
+
+def add_participants_to_conversation(
+    participant_emails: List[str], case: Case, db_session: SessionLocal
+):
+    """Adds one or more participants to the case conversation."""
+    if not case.conversation:
+        log.warning(
+            "Case participant(s) not added to conversation. No conversation available for this case."
+        )
+    plugin = plugin_service.get_active_instance(
+        db_session=db_session, project_id=case.project.id, plugin_type="conversation"
+    )
+    if not plugin:
+        log.warning(
+            "Incident participant(s) not added to conversation. No conversation plugin enabled."
+        )
+        return
+
+    try:
+        plugin.instance.add_to_thread(
+            case.conversation.channel_id, case.conversation.thread_id, participant_emails
+        )
+    except Exception as e:
+        event_service.log_case_event(
+            db_session=db_session,
+            source="Dispatch Core App",
+            description=f"Adding participant(s) to case conversation failed. Reason: {e}",
+            case_id=case.id,
+        )
+        log.exception(e)
+
+
+@background_task
+def case_add_or_reactive_participant_flow(
+    user_email: str,
+    case_id: int,
+    participant_role: ParticipantRoleType = ParticipantRoleType.participant,
+    service_id: int = 0,
+    event: dict = None,
+    organization_slug: str = None,
+    db_session=None,
+):
+    """Runs the case add or reactive participant flow."""
+    case = case_service.get(db_session=db_session, case_id=case_id)
+
+    if service_id:
+        # we need to ensure that we don't add another member of a service if one
+        # already exists (e.g. overlapping oncalls, we assume they will hand-off if necessary)
+        participant = participant_service.get_by_case_id_and_service_id(
+            case_id=case_id, service_id=service_id, db_session=db_session
+        )
+
+        if participant:
+            log.debug("Skipping resolved participant. Oncall service member already engaged.")
+            return
+
+    participant = participant_service.get_by_case_id_and_email(
+        db_session=db_session, case_id=case.id, email=user_email
+    )
+    if participant:
+        if participant.active_roles:
+            return participant
+
+        if case.status != CaseStatus.closed:
+            # we reactivate the participant
+            participant_flows.reactivate_participant(
+                user_email, case, db_session, service_id=service_id
+            )
+    else:
+        # we add the participant to the incident
+        participant = participant_flows.add_participant(
+            user_email, case, db_session, service_id=service_id, role=participant_role
+        )
+
+    # we add the participant to the tactical group
+    # add_participant_to_tactical_group(user_email, case, db_session)
```

Comment:
afaik cases don't use `user_middleware` because we don't enable commands. I supposed if we wanted to add more permissions around the modals that may be required in the future though.